### ISSUE TITLE
[Lake][I/O] Techspike - Middleware for parquet appending/mutating/other

### ISSUE THREAD

**Issue by idiom-bytes** [oceanprotocol/pdr-backend#572]
### Motivation
Parquet files work w/ partitioning, but you need to have a way about it. As we work towards building an incremental pipeline, we need an approach to append records rather than what we have today. Today, we simply append data and yield a new parquet file. As data grows, this i/o will become more expensive.

### Towards a solution
Rather than creating new records, parquet/polars supports partitions. This works very similarly as a regular .parquet file, except you have `file_partitions.parquet` folder, w/ many files, that you can read natively by polars by pointing at the folder name and treating it like a regular file. Just like any other db, your partioning can impact performance.

To build the partitions effectively, perhaps we should look at some thin middleware like fastparquet or duckdb to handle the partitioning/bucketing.

### Potential solutions
As we look to support enough tables + size on-disk, we'll need a solution sooner rather than later to avoid long disk I/O times.

First, **Apache Arrow** which is the standard for how polars is built, supports a broad approach to partitioning as described below. By using [pyarrow.parquet.write_to_dataset](https://arrow.apache.org/docs/python/generated/pyarrow.parquet.write_to_dataset.html)

![image](https://github.com/oceanprotocol/pdr-backend/assets/69865342/8172f1bc-ceb2-48c7-acd6-d8323a59ac3a)

https://arrow.apache.org/cookbook/py/io.html#writing-partitioned-datasets
https://arrow.apache.org/docs/python/generated/pyarrow.parquet.write_to_dataset.html
https://stackoverflow.com/questions/76121937/how-to-append-new-data-to-an-existing-parquet-file

Second, **Fastparquet** append/write w/ partitions: https://fastparquet.readthedocs.io/en/latest/quickstart.html#writing
Polars read parquet w/ partitions: https://docs.pola.rs/py-polars/html/reference/api/polars.read_parquet.html

Another potential solution would be to use **Duckdb**. However, as far as I understand it, we might need to first save the data to duckdb (db in-memory, persistent storage on analytics.db file). Such that we append to DuckDB, and then write new parquet files. DuckDB looks very promising, but leaves gaps as to whether it will allow us to append/grow incrementally.

Duckdb partitioned writes also seem to implement append in the same as pyarrow.parquet (using overwrite rules). **Please techspike to test/verify.**

https://duckdb.org/docs/data/partitioning/hive_partitioning
https://duckdb.org/docs/data/partitioning/partitioned_writes#overwriting

### Near-parquet
Polars supports other data-formats and processes (like Delta Lake) that may be able to help us across all 2 stages of the pipeline.
1. Delta Lake - Upstream/Fetching. Where we need to join + mutate records. I.e. Database like, to yield bronze-tables.
2. Arrow/Duck - Downstream/Computing. Where we just need to process records incrementally and can append-to-end/rebuild, to yield silver (aggregate) and gold (summary) tables.

### DoD
- [ ] Rather than creating new parquet files as a way to append, implement partitioning, in such a way that is saves files in efficient chunks.
- [ ] `gql_data_factory_save_parquet()` should be updated to benefit from append + partitioning
- [ ] `bronze_tables` should only save records changed, not entire partitions.
- [ ] `ohlcv_factory.save_rawohlcv_file` and any other parquet-component, should be updated to benefit from append + partitioning

**Comment by idiom-bytes** [comment]
Ok, so I found a really interesting approach that uses a vector db written in rust that uses the arrow format as the base layer.

This gives you the ability to work on a columnar file format, with everything on-disk, in-process, that handles appending/upserts/etc... and because it's all arrow in the base layer, you get all the serialization benefits, and integration w/ the data ecosystem.

It does this by using an inverted indexer which is different than how everyone else is doing things... This specific architecture unlocks what we're looking for here.
https://lancedb.github.io/lancedb/concepts/index_ivfpq/

Lancedb looks promising as it's all in-process, so there are no extra servers to run or infra (like competing spark solutions)

https://lancedb.github.io/lancedb/#open-source-and-cloud-solutions

It even has a pandas hook to retrieve a pd df... Which means we might be able to easily extend the interface to return a polars lazy frame, and continue to basically use a polars/arrow architecture which is serverless, on disk.

Polars support (and more) in the works.

**Comment by idiom-bytes** [comment]
We may need to consider some limitations of the limitations for how lance handles polars, such we can improve the performance.

TLDR; "We use the scan_pyarrow_dataset to convert to LazyFrame. Polars is able to push-down some filters to a pyarrow dataset, but the pyarrow dataset expects pyarrow compute expressions while Lance expects SQL strings. This means that we’ve had to disable the filter push-downs, meaning that Polars won’t be able to take advantage of the fast filtering and indexing features of Lance."

How:
(1) before doing the polars work, you fast filter/index w/ lance-friendly-sql
(2) get the results from (1) as a polars lazy frame that you then work with

**Comment by kdetry** [comment]
I have spent some time to look around with techs, and my consideration is below.

### Database Technologies

#### 1. **DuckDB**
- **Advantages**:
  - Embedded: Runs within your application, no need for a separate server.
  - Easy to set up and use, great for development and testing. (We can reach directly to data and we do not need an extra testing setting)
  - Good performance for analytical queries on moderate-sized datasets.
- **Disadvantages**:
  - Limited scalability for very large datasets or high-concurrency scenarios compared to distributed databases.
  - Primarily designed for single-node setups.

#### 2. **ClickHouse**
- **Advantages**:
  - High performance for analytical queries, even on very large datasets.
  - Scalable, can handle petabytes of data with clusters.
  - Supports real-time query processing and  SQL features.
- **Disadvantages**:
  - More complex to set up and manage than embedded databases. (We will need DevOps operations for testing)
  - Requires more resources (dedicated servers).

#### 3. **LanceDB**
- **Advantages**:
  - Specialized for handling and querying vector data, I can not comment about it because I still think the data that developed for is not our case.
  - Suitable for search engines and uses vectoral nodes
  - Efficient similarity search for high-dimensional vectors.
- **Disadvantages**:
  - Being a newer technology, it might have a smaller community and fewer integrations compared to more established databases.
  - Focused on vector data, which may be a limitation (I feel like our application requires extensive relational data processing)

### 4. **CouchDB**
- **Advantages**:
  - Ease of Use, has a HTTP API 
  - CouchDB is a NoSQL database, allowing for flexible schema design, which is great for applications with evolving data models.
  - scalable
- **Disadvantages**:
  - it may not be as efficient as other databases designed explicitly for analytical workloads on columnar data.
  - Not optimized for fast analytical queries on large datasets, especially compared to columnar databases.

### 5. **MariaDB ColumnStore**
- **Advantages**:
  - Offers a columnar storage engine 
  - Probably it has the largest SQL features among options
  - It is MySQL at the end of the day
- **Disadvantages**:
  - Setup and Management
  - Relative Performance.

### 6. **Apache Kudu**
- **Advantages**:
  - Fast Analytics on Fast Data
  - **Columnar Storage**
  - **Integration**: Integrates well with the Hadoop ecosystem, including tools like Apache Spark, Impala, and others for complex data processing pipelines.
- **Disadvantages**:
  - **Operational Complexity**
  - I guess no one has a long experience with it


### File Formats

#### 1. **Parquet**
- **Advantages**:
  - Efficient columnar storage format, great for analytical workloads.
  - Excellent compression and encoding schemes, reducing storage costs.
  - Broadly supported by data processing tools and cloud platforms.
- **Disadvantages**:
  - Primarily read-optimized, making updates and writes more complex.

#### 2. **ORC**
- **Advantages**:
  - Similar to Parquet in being an efficient columnar format with strong compression.
  - Offers fine-grained indexing, which can improve query performance.
- **Disadvantages**:
  - Like Parquet, more suited to read-heavy workloads.

#### 3. **CSV or JSON** 
- **Advantages**:
  - Simple and widely supported, making it great for data interchange.
  - Easy to read and write with most programming languages and tools.
- **Disadvantages**:
  - No built-in compression or optimization, leading to larger file sizes and slower queries on large datasets.
  - Again, slow queries

### Considerations for Scalability
- **ClickHouse** is the most scalable (if we do not count Kudu) option among the databases discussed, suitable for growing data needs and high query volumes.
- **DuckDB** offers convenient embedded analytics for mid-scale data or as a component of larger systems where distributed processing isn't needed.
- **LanceDB**is specialized for vector and AI-related workloads, but I guess the data model that we consider is not suitable for this one.
- For file formats, **Parquet** and **ORC** are more scalable in terms of storage efficiency and query performance on large datasets

**Comment by idiom-bytes** [comment]
I'm looking specifically for an arrow-based approach that has polars-native integration such that we can do e2e testing on top of simple, on-disk data, without having to resort to integrating w/ a local db.

I'm also looking for an in-process approach such that we don't have to pass down large memory and compute requirements to our users such that they require a cluster off-the-bat in order to support a basic filesystem + queries.

Clickhouse + other DBs require multiple servers to keep the FS + querying available.
Arrow-based systems run on-top off S3. They only need whatever is querying to run the query. It's serverless.

**This passes a much smaller set of requirements and scope down to our users.**

LanceDB is built on-top of the arrow, columnar-based filesystem, **as a base-level implementation it's a columnar database**

The discussion is of whether we want to continue focus on a serverless, in-process, arrow-based system, that supports all the needs we have and can solve our problems at-scale, rather than a cluster-based approach.

**Comment by idiom-bytes** [comment]
**[Further on Lance dovetailing with AI]**
Am also now realizing the benefits that a vector db w/ good random sampling can do for the ML and MPP, and how one doevetailing the other more and more...

**[Reading from filesytem for training]**
Example... downstream, spark/cluster/data will have to read/get data to test the model `sample(x_train,y_train)`, which is exactly what a vectordb is designed to do efficiently. For other db models (sql, couch), we'll either need an inefficient connector and because there are many things hitting the cluster, more and more provisioning.
![image](https://github.com/oceanprotocol/pdr-backend/assets/69865342/c3944ec8-3d76-4374-b038-51b0fd3f8659)

With lancedb, provisioning is done at the process level (i.e. every predictoor reading from disk pays their part) rather than at the cluster level (shared service).

**[Distributed computing and MPP]**
Due to arrow/provisioning for ML workloads, it's incredibly easy to sample/shuffle data for distributed workloads. Other dbs like redshift/etc... do not have this. All of this data also needs to exist on disk (i.e. S3) so it can be sampled/shuffled by the system.... this is one of those ops + infra + serialization steps that takes a ton of time to get implemented and costs a ton of time + $$$ every time you run.

https://lancedb.github.io/lance/integrations/tensorflow.html#distributed-training-and-shufflingexanpo

So again, things that look like will be solved now, I believe will lead to a ton of hurdles and costs further down the timeline.

**Comment by idiom-bytes** [comment]
Research and tech spikes delivered.
Decision is to build out infra w/ duckdb.
